test/test_span_token.py

```diff
@@ -53,6 +53,14 @@ def test_parse_in_emphasis(self):
     def test_parse_in_strikethrough(self):
         self._test_parse_enclosed(span_token.Strikethrough, '~~')
 
+    def test_preserve_space(self):
+        self._test_parse(span_token.InlineCode, '``` ```', ' ')
+        self._test_parse(span_token.InlineCode, '`  ``  `', '  ``  ')
+
+    def test_preserve_escapes(self):
+        self._test_parse(span_token.InlineCode, '`\\xa0b\\xa0`', '\\xa0b\\xa0')
```

Comment:
> Solved by converting line endings to space and preserving space and escape sequences in InlineCode tokens

OK, so this one tests that escape sequences are preserved - but do I see it correctly, that regarding _escape sequences_, no change was necessary / done?

Reply:
That's right, there is a new test case to check that escape sequences are preserved according to the spec, but no change to the logic was required.

Reply:
Great, thanks for the thorough review!

1. I understand your point of wanting to make minimal changes and keep logic where it is. However, in this case, I'd like to ask you to reconsider -- let me explain why! Currently, the token classes do the parsing and some processing, and the renderer classes process the input, too. It would be helpful if the token classes would limit themselves to only do the parsing, and leave all processing to the renderers -- because different renderers may have different needs. I'm working on a Markdown renderer, for example, where the goal is to be able to do a near-perfect round trip. That's not possible if the source has been processed in an irreversible way. (I am nearly finished with a first iteration of that renderer, and it does have a few more little source-preserving changes like this one.) I am of course fully aware that mistletoe was not written with "my" use case in mind, but I do believe that a clear separation of concerns between parsing and rendering can benefit the project as a whole anyway.

3. Correct. The classic off-by-one.

4. If you try and run the 0.30 tests on Windows, you'll probably see the problem. One of the new, failing test cases emits an invalid Unicode codepoint and then the test fails with an exception because of its own printout. It's for sure not an essential patch; I just like tests to run cleanly.

Reply:
Thanks for your feedback, here are my remarks to the 1st point:

Generally, I agree with you, parsing and rendering should be as separated as possible. In this concrete case though, I still believe that this new handling of extra spaces around the inline code itself should be considered as related just to parsing.

Yes, every single renderer could decide itself what to do with those extra spaces, but I don't see much added value here. I think it would be similar to leaving renderers decide whether they want to unescape entity references, or not (unescaping is done on the parsing level since #135).

Of course, the renderer _to Markdown_ which you prepare would most probably benefit from knowing that extra spaces were used, so it can preserve the original spacing. Yet isn't this a special case of a renderer? I expect that you already add some extra info to the parsed tokens, so I can imagine that storing an additional flag of "extra spaces present" could do the job?

Reply:
Ad 3. I've checked again and I still can't see the "invalid Unicode codepoint" problem - just 11 failed spec tests (running `py -m test.specification` with v0.30 of CommonMark downloaded). So probably yes, for a better focus on just that problem, I would suggest to move it into an extra PR. Thanks in advance.

PS: I've just noticed the last 3 comments should be moved 1 level up (out of code line conversation), but whatever, next time... :)

Reply:
Ad 1. There is also this stuff of _replacing new lines with space_ within inline code - I think that this is also something that authors of the CommonMark spec consider "getting rid of input formatting which is not meant to be presented by renderers". I hope it makes sense now why I'm against keeping this logic on the renderers' part, I hope it wasn't too verbose... :)

Reply:
Sure, though I doubt that the authors of the CommonMark spec care much about other renderers/transformations than Markdown to HTML. 😃

Reply:
OK, thanks for the amendments.

Sure, the primary output format is definitely HTML, so it is only logical that the spec authors illustrate the parsing results on the rendered HTML. It's hard for me to tell how much they actually care about the other output formats, but I feel the spec is written in a way that it basically focuses on how to interpret the input markdown, getting an AST as a result, and that's it... :)

Reply:
I can just add that my understanding is that the spec only describes how to get from Markdown to HTML. The AST is an implementation detail not covered by the spec, where Markdown-to-HTML processors have full artistic freedom. But I may be wrong. And of course, AST users are probably not too happy about breaking changes to the API.

Reply:
Yeah, it looks we see it a little bit differently, partly meaning the same thing, but using different words for it (like when I agree with you a concrete AST really is an implementation detail of a given parser), but I think that never mind. :) BTW Good point about a possible "breaking change", I will try to mention it in the next release notes somehow...